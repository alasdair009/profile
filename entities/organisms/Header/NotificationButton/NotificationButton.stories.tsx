import { Meta, StoryObj } from "@storybook/nextjs";
import { NotificationButton } from "./NotificationButton";

const meta: Meta<typeof NotificationButton> = {
  component: NotificationButton,
  args: {
    isSubscribed: false,
  },
};
export default meta;

export const Default: StoryObj<typeof NotificationButton> = {};
