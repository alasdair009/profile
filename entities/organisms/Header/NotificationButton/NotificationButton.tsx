"use client";

import notificationIcon from "@/entities/assets/notification.svg";
import { NotificationButton as Root, NotificationIcon } from "../styles";
import { HTMLAttributes, useCallback, useEffect, useState } from "react";
import useFcmToken from "@/lib/notifications/useFCMToken";
import useFCMToken from "@/lib/notifications/useFCMToken";
import { getMessaging, getToken } from "firebase/messaging";
import firebaseApp from "@/firebase";

type NotificationButtonProps = {
  isSubscribed: boolean;
} & HTMLAttributes<HTMLButtonElement>;

export function NotificationButton({
  isSubscribed,
  ...rest
}: NotificationButtonProps) {
  const [notificationState, setNotificationState] =
    useState<NotificationPermission>("denied");
  const [fcmToken, setFcmToken] = useState("");
  const [notificationPermissionStatus, setNotificationPermissionStatus] =
    useState("");

  const permissionRequest = () => {
    const retrieveToken = async () => {
      try {
        if (typeof window !== "undefined" && "serviceWorker" in navigator) {
          const messaging = getMessaging(firebaseApp);

          // Request notification permission
          const permission = await Notification.requestPermission();
          setNotificationPermissionStatus(permission);

          if (permission === "granted") {
            const currentToken = await getToken(messaging, {
              vapidKey:
                "BCGeeUt-hVXpwPelsOIsdohGlCXmxpK8VjkuEA3z6zGzyFLDt35dQm_pAfgV3ylRI1gntXKsIDdMaDQXmR8tbMg", // Replace with your Firebase project's VAPID key
            });
            if (currentToken) {
              setFcmToken(currentToken);
            } else {
              console.log(
                "No registration token available. Request permission to generate one."
              );
            }
          }
        }
      } catch (error) {
        console.log("Error retrieving token:", error);
      }
    };

    retrieveToken();
  };

  return (
    <Root
      type="button"
      variant="transparent"
      onClick={() => permissionRequest()}
      {...rest}
    >
      <NotificationIcon
        src={notificationIcon}
        alt="Notifications icon"
        $isSubscribed={isSubscribed}
        style={{
          background: notificationState === "granted" ? "green" : "red",
          // display: "none",
        }}
        data-sub={notificationPermissionStatus}
      />
    </Root>
  );
}
