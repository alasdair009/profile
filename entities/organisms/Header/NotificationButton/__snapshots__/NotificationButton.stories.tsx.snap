// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`organisms/Header/NotificationButton Default smoke-test 1`] = `
<button data-testid="Button"
        type="button"
        class="sc-iCZwEW hJjhSG sc-drnuxz"
>
  <img alt="Notifications icon"
       loading="lazy"
       width="800"
       height="800"
       decoding="async"
       data-nimg="1"
       class="sc-hDcvty kKFXDJ"
       srcset="/static/media/entities/assets/notification.svg?w=828&amp;q=75 1x, /static/media/entities/assets/notification.svg?w=1920&amp;q=75 2x"
       src="/static/media/entities/assets/notification.svg?w=1920&amp;q=75"
       style="color: transparent; background: red; display: none;"
  >
</button>
`;
